source/common/common/posix/thread_impl.cc

```diff
@@ -27,6 +27,11 @@ int64_t getCurrentThreadId() {
 #endif
 }
 
+ThreadId getCurrentThreadId() {
+  static thread_local ThreadId cached_thread_id = ThreadId(getCurrentThreadIdBase());
```

Comment:
I think this is a destructor-race fiasco and you need to make this a pointer -- see CONSTRUCT_ON_FIRST_USE -- actually you might as well just use that or make a thread_local variant of it.

Reply:
OKay, I can do that. But I still have a quesion:

> I think this is a destructor-race fiasco and you need to make this a pointer

Hmmm, How can that happen? This is a private method that could only be called by the `currentThreadId()` of ThreadFactory. That means no static variable could reference it directly.

Reply:
It even get minor performance improvement (from 2.3 ns -> 2.1 ns). Although I don't know why. The new implementation should add additional de-referencing to pointer.

```
bmGetCurrentThreadId/0       2.09 ns         2.09 ns    336280476
bmGetCurrentThreadId/1       2.10 ns         2.10 ns    338040767
bmGetCurrentThreadId/2       2.09 ns         2.09 ns    337318151
```

Reply:
PTAL @ https://www.freecodecamp.org/news/cpp-static-initialization-order-fiasco/

That explains the destruction-order fiasco

Reply:
currentThreadId() can be called during destruction of some other object in a different translation unit.

Reply:
Yeah I have read the doc long time ago. I mean the destruction-order fiasco happens when the another static variable try to call the `currentThreadId()` when it doing the destruction.

Hah, I see, if someone create a static instance of `ThreadFactory` and call the `currentThreadId()` by the static `ThreadFactory` instance, then it's possible.